src/tools/sssctl/sssctl_logs.c

```diff
@@ -212,6 +381,57 @@ errno_t get_confdb_sections(TALLOC_CTX *ctx, struct confdb_ctx *confdb,
     return ret;
 }
 
+const char **get_targets(TALLOC_CTX *ctx,
+                         int services_mask, const char **domainv)
+{
+    int i;
+    int count = 1;
+    const char **targets = NULL;
+
+
+    targets = talloc_zero_array(ctx, const char *, count);
+    if (targets == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Could not allocate memory for the list of targets\n");
+        return NULL;
+    }
+
+    if (services_mask != 0) {
+        for (i = 0; i < SERV_COUNT; i++) {
+            if (services_mask == 0 || (services_mask & services[i].mask) != 0) {
+                targets = talloc_realloc(ctx, targets, const char *, count + 1);
+                if (targets == NULL) {
+                    DEBUG(SSSDBG_CRIT_FAILURE,
+                          "Could not allocate memory for the list of targets\n");
+                    return NULL;
+                }
+                targets[count - 1] = talloc_strdup(ctx, services[i].name);
+                targets[count++] = NULL;
+            }
+        }
+    }
+
+    if (domainv != NULL) {
+        for (; *domainv != NULL; domainv++) {
+            targets = talloc_realloc(ctx, targets, const char *, count + 1);
+            if (targets == NULL) {
+                DEBUG(SSSDBG_CRIT_FAILURE,
+                      "Could not allocate memory for the list of targets\n");
+                return NULL;
+            }
+            if (STARTS_WITH(*domainv, "domain/")) {
+                targets[count - 1] = talloc_strdup(ctx, *domainv);
+            } else {
+                targets[count - 1] = talloc_asprintf(ctx, "domain/%s", *domainv);
+            }
+            targets[count++] = NULL;
+        }
+    }
+
+
```

Comment:
Clean up on error.

Reply:
Implemented a temporay memory context for this function.